src/compiler/checker.ts

```diff
@@ -22486,7 +22486,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 /*stringsOnly*/ false,
                 t => void mappedKeys.push(instantiateType(nameType, appendTypeMapping(targetType.mapper, getTypeParameterFromMappedType(targetType), t))),
             );
-            return getUnionType(mappedKeys);
+            return mappedKeys.length ? getUnionType(mappedKeys) : stringNumberSymbolType;

```

Comment:
On second thought, this likely should only happen when obtaining apparent mapped type keys on the source side of the relation check. I'll think about the test case for the target side that fails with this.

The problem here is that when `modifiersType` is `unknown` (or `{}`) then `keyof modifiersType` is `never`. Since `never` is assignable to everything, the checker assumed that `keyof { [P in keyof T as T[P] extends string ? P: never]: any }` is assignable to `string`. 

Then based on that information it managed to simplify the instantiated substitution type of `Extract` to just `keyof { [P in keyof T as T[P] extends string ? P: never]: any }` (its `newBaseType`). Then in turn `getNarrowableTypeForReference` concluded that it should substitute constraints. Those were substituted with `stringNumberSymbolType` (for the `keyof ...`) and that's not assignable to `string`.

Before https://github.com/microsoft/TypeScript/pull/56742 , it didn't mistakenly assume that `keyof { [P in keyof T as T[P] extends string ? P: never]: any }` is assignable to `string` so the substitution type was instantiated as `(keyof { [P in keyof T as T[P] extends string ? P: never]: any }) & string` and that is assignable to a string.